onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc

```diff
@@ -0,0 +1,508 @@
+      }
+      std::vector<int64_t> output_shape;
+      const auto& output_defs = node.OutputDefs();
+      ORT_RETURN_IF_NOT(GetShape(*output_defs[0], output_shape, logger), "Cannot get shape");
+      for (uint32_t axis = 0, axis_count = static_cast<uint32_t>(output_shape.size()); axis < axis_count; ++axis)
+      {
+          if (~kept_axes_mask & (1<<axis))
+          {
+              reduced_axes.push_back(axis);
+          }
+      }
+
+      emscripten::val input = model_builder.GetOperand(node.InputDefs()[0]->Name());
+      emscripten::val options = emscripten::val::object();
+      options.set("keepDimensions", false);
+
+      std::vector<int64_t> input_shape;
+      ORT_RETURN_IF_NOT(GetShape(*input_defs[0], input_shape, logger), "Cannot get shape");
+      const auto input_rank = input_shape.size();
+
+      std::vector<int32_t> axes_data;
+      std::transform(
+          reduced_axes.begin(), reduced_axes.end(), std::back_inserter(axes_data),
+          [input_rank](int64_t axis) -> int32_t { return SafeInt<int32_t>(HandleNegativeAxis(axis, input_rank)); });
+      options.set("axes", emscripten::val::array(axes_data));
+
+      output = model_builder.GetBuilder().call<emscripten::val>("reduceSum", input, options);
+    }
+    break;
+
+  case RecognizedOperatorType::Transpose:
+    {
+      emscripten::val input = model_builder.GetOperand(node.InputDefs()[0]->Name());
+      // Transpose via input strides. The output tensor is not strided.
+      assert(m_components.front().GetDimensionCount() == m_components.back().GetDimensionCount());
+      // Remap transposed strides using the component labels from input to output.
+      auto label_indices = m_components.back().GetLabels(m_label_indices);
+
+      std::vector<uint32_t> permutation{label_indices.begin(), label_indices.end()};
+      emscripten::val options = emscripten::val::object();
+      options.set("permutation", emscripten::val::array(permutation));
+      output = model_builder.GetBuilder().call<emscripten::val>("transpose", input, options);
+    }
+    break;
+  case RecognizedOperatorType::Identity:
+    {
+      emscripten::val input = model_builder.GetOperand(node.InputDefs()[0]->Name());
+      output = model_builder.GetBuilder().call<emscripten::val>("identity", input);
+    }
+    break;
+  default:
+    break;
+  }
+
+  model_builder.AddOperand(node.OutputDefs()[0]->Name(), std::move(output));
+  return Status::OK();
+}
+
+// Operator support related.
+
+bool EinsumOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers,
+                                        const Node& node,
+                                        const WebnnDeviceType device_type,
+                                        const logging::Logger& logger) const {
+
+  if (device_type == WebnnDeviceType::CPU) {
+    LOGS(logger, VERBOSE) << "Einsum is not supported for cpu in WebNN EP. Matmul and ReduceSum are not supported in XNNPACK.";
```

Comment:
`Matmul` is just implemented in Chromium, we should only restrict `recognized_operator_type == RecognizedOperatorType::ReduceSum`